File: DPS_DTCL/IspProtocol/IspSubCommandProcessor.cs
using DTCL.Log;
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices.WindowsRuntime;

namespace IspProtocol
{
    public class IspSubCommandProcessor
    {
        readonly Dictionary<byte, IIspSubCommandHandler> handlers = new Dictionary<byte, IIspSubCommandHandler>();

        /// <summary>
        /// Registers a handler for a given subcommand.
        /// </summary>
        public void Register(byte subCmd, IIspSubCommandHandler handler)
        {
            // Check if already registered with same handler
            if (handlers.TryGetValue(subCmd, out var existingHandler) && existingHandler == handler)
            {
                return; // Already registered, do nothing
            }
            
            handlers[subCmd] = handler;
            Log.Info($"[EVT3001] Registered handler for subcommand 0x{subCmd:X2}.");
        }

        /// <summary>
        /// Invokes the receive data processor for a given subcommand.
        /// </summary>
        public uint ProcessRxSubCommand(byte subCmd, byte[] data)
        {
            if (handlers.TryGetValue(subCmd, out var handler))
            {
                Log.Info($"[EVT3002] Processing RX data for subcommand 0x{subCmd:X2}, Size={data?.Length ?? 0} bytes.");
                var res = handler.processRxData(data, subCmd);
                return res;
            }
            else
            {
                Log.Warning($"[EVT3003] No RX handler registered for subcommand 0x{subCmd:X2}.");
                return 1;
            }
        }

        /// <summary>
        /// Prepares the processor for receiving expected RX data.
        /// </summary>
        public long prepareForRx(byte subCmd, byte[] data)
        {
            if (handlers.TryGetValue(subCmd, out var handler))
            {
                Log.Info($"[EVT3004] prepareForRx called for subcommand 0x{subCmd:X2}, ExpectedSize={data?.Length ?? 0}.");
                return handler.prepareForRx(data, subCmd, data.Length);
            }
            else
            {
                Log.Warning($"[EVT3005] No prepareForRx handler registered for subcommand 0x{subCmd:X2}.");
                return 1024; // fallback default
            }
        }

        /// <summary>
        /// Retrieves TX data by invoking the appropriate subcommand handler.
        /// </summary>
        public byte[] PrepareTxData(byte subCmd, byte[] data)
        {
            if (handlers.TryGetValue(subCmd, out var handler))
            {
                Log.Info($"[EVT3006] Preparing TX data for subcommand 0x{subCmd:X2}, InputSize={data?.Length ?? 0}.");
                return handler.prepareDataToTx(data, subCmd);
            }
            else
            {
                Log.Warning($"[EVT3007] No TX handler registered for subcommand 0x{subCmd:X2}. Returning empty.");
                return Array.Empty<byte>();
            }
        }
    }
}

File: Firmware/DTCL/Core/Src/Protocol/IspSubCommandProcessor.cpp
#include "IspSubCommandProcessor.h"
#include <cstdio>

IspSubCommandProcessor::IspSubCommandProcessor() : handlerCount(0) {
    // Initialize array
    for (uint8_t i = 0; i < MAX_HANDLERS; i++) {
        handlers[i].valid = false;
    }
}

void IspSubCommandProcessor::registerHandler(uint8_t subCmd, IIspSubCommandHandler* handler) {
    if (handlerCount < MAX_HANDLERS) {
        handlers[handlerCount].subCmd = subCmd;
        handlers[handlerCount].handler = handler;
        handlers[handlerCount].valid = true;
        handlerCount++;
    }
}

IIspSubCommandHandler* IspSubCommandProcessor::findHandler(uint8_t subCmd) {
    for (uint8_t i = 0; i < handlerCount; i++) {
        if (handlers[i].valid && handlers[i].subCmd == subCmd) {
            return handlers[i].handler;
        }
    }
    return nullptr;
}

uint32_t IspSubCommandProcessor::prepareForRx(uint8_t subCmd, const uint8_t* data, uint32_t len) {
    IIspSubCommandHandler* handler = findHandler(subCmd);
    if (handler) {
        return handler->prepareForRx(data, subCmd, len);
    } else {
        printf("[RX] No handler for sub command 0x%02X\n", subCmd);
    }
    return 1024; //default
}

uint8_t IspSubCommandProcessor::processRxSubCommand(uint8_t subCmd, const uint8_t* data, uint32_t len) {
    IIspSubCommandHandler* handler = findHandler(subCmd);
    if (handler) {
        handler->processRxData(data, subCmd, len);
    } else {
        printf("[RX] No handler for sub command 0x%02X\n", subCmd);
    }
    return 0;
}

uint8_t IspSubCommandProcessor::prepareTxData(uint8_t subCmd, const uint8_t* data, uint32_t& outLen) {
    IIspSubCommandHandler* handler = findHandler(subCmd);
    if (handler) {
        return handler->prepareDataToTx(data, subCmd, outLen);
    }
    printf("[TX] No handler for sub command 0x%02X\n", subCmd);
    outLen = 0;
    return 1;
}
